### ISSUE TITLE
Add a review/audit property to a Book/Chapter/Page

### ISSUE THREAD

**Issue by mikeyz24** [BookStackApp/BookStack#2123]
**Describe the feature you'd like**
I'm using Bookstack for documentation and policies & procedures purposes.  These items need to be reviewed regularly so that they are kept up to date.  Adding a review date to a Book/Chapter/Page along with email notifications to the owner of upcoming items for review would be very useful.  A history of when a review was completed and by whom should also exist.

**Describe the benefits this feature would bring to BookStack users**
It would allow people to keep their documentation up-to-date and keep an audit trail for accountability purposes.

**Additional context**

**Comment by tiredofit** [comment]
This would be quite an interesting feature to add introducing review/moderation workflow. I envision another set of permissions on the Shelf level and even Book level that allows a certain set of permissions to Create/Edit/Update content, yet keep in a "Review" mode similar to Draft which notifies in the Application a series of pages/content to be reviewed. The "Reviewer" role would then do any edits required and be responsible for "Publishing".

**Comment by BFjimmy** [comment]
<span lang="EN-US" style="mso-ansi-language: EN-US;">After analyzing the database structure and current functionality, here’s a proposal for introducing a review workflow in BookStack to support a "peer review" process before publishing pages.</span>

**Database Observations**

**Draft Pages**:

- <span lang="EN-US" style="mso-ansi-language: EN-US;">A page is marked as a draft if the draft column in the pages table is set to 1. This column appears to only be used before a page is first published. </span>Once published, it is no longer utilized.
- <span lang="EN-US" style="mso-ansi-language: EN-US;">Subsequent unpublished changes are tracked using the page\_revisions table with a type value of update\_draft.</span>

**Concurrent Drafts**:

- <span lang="EN-US" style="mso-ansi-language: EN-US;">Multiple users can work on drafts simultaneously, which may cause changes by one user to be hidden from another. This is acceptable due to the notification system alerting users to such conflicts.</span>

**<span lang="EN-US" style="mso-ansi-language: EN-US;">Proposed Review Process</span>**

<span lang="EN-US" style="mso-ansi-language: EN-US;">To introduce a peer review mechanism, the following changes and behaviors are suggested:</span>

**Button Behavior Changes**:

<span lang="EN-US" style="mso-ansi-language: EN-US;">For users without the publish page permission:</span>
- <span lang="EN-US" style="mso-ansi-language: EN-US;">Replace the "Save Page" button with a "Set to Review" button. The "Save Draft" button remains unchanged.</span>

<span lang="EN-US" style="mso-ansi-language: EN-US;">For users with publish page permissions:</span>
- <span lang="EN-US" style="mso-ansi-language: EN-US;">Retain both the "Save Page" and "Set to Review" options, allowing these users to optionally participate in the review process.</span>

**<span lang="EN-US" style="mso-ansi-language: EN-US;">Database Updates for Review Status</span>**<span lang="EN-US" style="mso-ansi-language: EN-US;">:</span>

- <span lang="EN-US" style="mso-ansi-language: EN-US;">When a user clicks "Set to Review," update the type column in the page\_revisions table for that revision to review\_draft.</span>

**Behavior for Reviewing Drafts**:

<span lang="EN-US" style="mso-ansi-language: EN-US;">Any user with the publish page permission:</span>
- <span lang="EN-US" style="mso-ansi-language: EN-US;">Can load the "review draft" of another user as their own draft but **cannot make changes**.</span>
- <span lang="EN-US" style="mso-ansi-language: EN-US;">Gains the option to publish the page directly based on the review draft.</span>

<span lang="EN-US" style="mso-ansi-language: EN-US;">If a page is marked as type = review\_draft and the draft owner revisits it:</span>
- <span lang="EN-US" style="mso-ansi-language: EN-US;">They can only choose to "Remove from Review," reverting the draft to an editable state.</span>

**Restrictions**:

<span lang="EN-US" style="mso-ansi-language: EN-US;">While a page is in review\_draft status:</span>
- <span lang="EN-US" style="mso-ansi-language: EN-US;">Only the draft owner can remove it from the review queue.</span>
- <span lang="EN-US" style="mso-ansi-language: EN-US;">Users with publish page permissions can only publish it but not edit it.</span>

**Intended Workflow**

<span lang="EN-US" style="mso-ansi-language: EN-US;">A user without publish page rights works on a draft and clicks "Set to Review" instead of "Save Page." </span>The draft's status changes to review\_draft.

<span lang="EN-US" style="mso-ansi-language: EN-US;">A user with publish page rights sees the page marked for review, loads the draft, and can decide to:</span>

- <span lang="EN-US" style="mso-ansi-language: EN-US;">Publish the draft as-is, or</span>
- <span lang="EN-US" style="mso-ansi-language: EN-US;">Notify the original author to make further changes by leaving the draft in the review state.</span>

<span lang="EN-US" style="mso-ansi-language: EN-US;">If the original author decides to remove the draft from the review queue, they regain the ability to edit it.</span>

**Advantages of This System**

**Simplicity**:

- <span lang="EN-US" style="mso-ansi-language: EN-US;">The system avoids implementing a complex review mechanism while providing a clear path for review and publication.</span>

**Seamless Integration**:

- <span lang="EN-US" style="mso-ansi-language: EN-US;">Leverages existing database structures and user permissions, minimizing changes to the core functionality.</span>

**Flexibility**:

- <span lang="EN-US" style="mso-ansi-language: EN-US;">Users with publish page permissions can participate in the review process or bypass it entirely, depending on the context.</span>

**Potential Implementation Steps**

1. <span lang="EN-US" style="mso-ansi-language: EN-US;">Add a new "Set to Review" button with associated logic.</span>
2. <span lang="EN-US" style="mso-ansi-language: EN-US;">Update the page\_revisions table to support the new review\_draft status.</span>
3. <span lang="EN-US" style="mso-ansi-language: EN-US;">Implement restrictions for users accessing review\_draft revisions.</span>
4. <span lang="EN-US" style="mso-ansi-language: EN-US;">Add UI indicators to highlight pages in the review process for users with publish page permissions.</span>

Also, maybe consider changing the text "Save page" to "Publish page" if the publish rights are enabled?

**Comment by BFjimmy** [comment]
@ssddanbrown Any thoughts on my suggestion?

**Comment by ssddanbrown** [comment]
Hi @BFjimmy, Thanks for offering up the proposal. This issue thread is really for post-publish review/auditing, whereas your proposal is more aligned to pre-publish review/auditing, which better fits #473.

Setting that aside, I like the simplicity of what's proposed but I think even that would still add considerably complexity once all edge-cases, interactions, added views and required controls have been fleshed out, and this is an area that will spawn a lot of additional requests & can be quite opinionated as it can be specific to company/business workflow. Implementation is commonly the lesser effort/concern/consideration, with general scope/complexity/maintenance/support being bigger factors which this will have significant impact on.

This does provide a good basis for an an initial implementation when we get to that point, but I'm not keen to jump into this further right now since I've recently already significant increased the scope with a couple of significant features (New editor and new [zip import/export format](https://github.com/BookStackApp/BookStack/pull/5260)) which I need to see settle somewhat to assess burden, and since I think I need to think through some of the higher level planning/ideas around draft collaboration/access in general, as that has a big impact to such a process/workflow. Since these areas are in high demand though, it may be something I return to and potentially focus on for 2025.

**Comment by BFjimmy** [comment]
Thanks for the reply @ssddanbrown

**Comment by brynmoorhouse** [comment]
Can I add my vote for this one please?

**Comment by nimble-software** [comment]
Agreed, this feature would be very useful

**Comment by kasdkuls** [comment]
Agreed, woud love to see this feature

**Comment by Grovkillen** [comment]
We're investigating in having this done using Redmine + a making a plugin. It's very specific since you would need to have Redmine alongside Bookstack but for our company this is a potentially perfect match.

**Comment by kevin39** [comment]
In case this helps someone: on my side, I implemented a small review process using only the tag system and search filters.

- We added a link in the header menu containing two predefined searches: “Waiting pages” and “Anomaly pages”.
- Each page includes a specific tag indicating its review state (waiting, verified, anomaly). The key point is that the author assigns the initial tag, and the reviewer updates it after verification.
- Each page also displays a dynamic warning (using JavaScript) whenever the tag value is anything other than “verified”.

<img width="1532" height="364" alt="Image" src="https://github.com/user-attachments/assets/e11e9f4e-ede1-499d-9cc6-923e2e8f955f" />

<img width="995" height="283" alt="Image" src="https://github.com/user-attachments/assets/de471576-2d76-4feb-a1f8-0cfb03bb6030" />

**Comment by balta1980** [comment]
Yes, this is just what we need in order to use it on our management system. Please.

**Comment by ssddanbrown** [comment]
BookStack is now [managed on Codeberg](https://codeberg.org/bookstack/bookstack).

As part of this change, we're closing all existing GitHub issues and pull requests.
You'll instead find this issue on Codeberg here: 
https://codeberg.org/bookstack/bookstack/issues/2123
  
Any further discussion for this will now be done on Codeberg.
Therefore this GitHub thread will be locked.
